site/specification/ORCv1.md

```diff
@@ -369,6 +397,194 @@ message Metadata {
+UNKNOWN_ENCRYPTION instead of a real value.
+
+```
+enum EncryptionAlgorithm {
+  // used for detecting future algorithms
+  UNKNOWN_ENCRYPTION = 0;
+  // 128 bit AES/CTR
+  AES_128 = 1;
+  // 256 bit AES/CTR
+  AES_256 = 2;
+}
+```
+
+## Encryption Variants
+
+Each encrypted column is written as two variants:
+
+* encrypted unmasked - for users with access to the key
+* unencrypted masked - for all other users
+
+The changes to the format were done so that old ORC readers will read
+the masked unencrypted data. Encryption variants encrypt a subtree of
+columns and use a single local key. The initial version of encryption
+support only allows the two variants, but this may be extended later
+and thus readers should use the first variant of a column that the
+reader has access to.
+
+```
+message EncryptionVariant {
+  // the column id of the root column that is encrypted in this variant
+  optional uint32 root = 1;
+  // the key that encrypted this variant
+  optional uint32 key = 2;
+  // the encrypted key for the file footer
+  optional bytes encryptedKey = 3;
+  // the stripe statistics for this variant
+  repeated Stream stripeStatistics = 4;
+  // encrypted file statistics as a FileStatistics
+  optional bytes fileStatistics = 5;
+}
+```
+
+Each variant stores stripe and file statistics separately. The file
+statistics are serialized as a FileStatistics, compressed, encrypted
+and stored in the EncryptionVariant.fileStatistics.
+
+```
+message FileStatistics {
+  repeated ColumnStatistics column = 1;
+}
+```
+
+The stripe statistics for each column are serialized as
+ColumnarStripeStatistics, compressed, encrypted and stored in a stream
+of kind STRIPE_STATISTICS. By making the column stripe statistics
+independent of each other, the reader only reads and parses the
+columns contained in the SARG.
+
+```
+message ColumnarStripeStatistics {
+  // one value for each stripe in the file
+  repeated ColumnStatistics colStats = 1;
+}
+```
+
+## Stream Encryption
+
+Our encryption is done using AES/CTR. CTR is a mode that has some very
+nice properties for us:
+
+* It is seeded so that identical data is encrypted differently.
+* It does not require padding the stream to the cipher length.
+* It allows readers to seek in to a stream.
+* The IV does not need to be randomly generated.
+
+To ensure that we don't reuse IV, we set the IV as:
+
+* bytes 0 to 2  - column id
+* bytes 3 to 4  - stream kind
+* bytes 5 to 7  - stripe id
+* bytes 8 to 15 - cipher block counter
+
+However, it is critical for CTR that we never reuse an initialization
+vector (IV) with the same local key.
+
+For data in the footer, use the number of stripes in the file as the
+stripe id. This guarantees when we write an intermediate footer in to
+a file that we don't use the same IV.
+
+Additionally, we never reuse a local key for new data. For example, when
+merging files, we don't reuse local key from the input files for the new
+file tail, but always generate a new local key.
```

Comment:
I wonder what would happen in append scenario. I think we will just rotate IV by increasing stripe id.

Reply:
The goal is to allow merging files without re-encrypting the data in the stripe. By putting the encryption stripe id and encrypted local keys in to the stripe information, the merge will do the right thing. (Assuming that the entire file is merged, since the writer is only setting the encryption information in the first stripe of each file. If someone builds a merge tool that merges partial files, they'll need to take care to set the properties on the transitions.)

Reply:
My question is specifically about appending scenario - you open an existing Orc file and append data at the tail. But I guess it would be very similar to merge case.

Reply:
Yes, it would be very similar to the merge case.